### PR TITLE
fix: upgrade phone rule to get only phone keyworld

### PR DESCRIPTION
**Pull request by luizleite-hotmart**
There's a small problem with the phone rule. In case the field is called phoneDetail will check correctly, but in case the field is just phone the regex would not get it.

### CODE REVIEW

**Comment by ojaswa1942** [comment]
@luizleite-hotmart Thanks for the contribution. 
@HiteshMah-Jan Adding you as a reviewer.

@luizleite-hotmart Can you please also open a corresponding issue that describes the problem this PR fixes? You can mention the PR in the issue too.

**Review by HiteshMah-Jan — CHANGES_REQUESTED** [review]

File: rules/sources/contact_data.yaml [hunk 1]
```diff
@@ -25,7 +25,7 @@ sources:
     isSensitive: False
     sensitivity: medium
     patterns:
-      - "(?i).*((phone|mobile|cellphone|contact)[^\\s/(;)#|,=!>]{0,2}(detail|details)|(?:mobile|phone|cellphone|contact)[^\\s/(;)#|,=!>]{0,10}(?:number|nbr|num|no)|(?:primary|secondary)[^\\s/(;)#|,=!>]{0,10}(phone|landline|telephone))"
+      - "(?i).*((phone|mobile|cellphone|contact)|(phone|mobile|cellphone|contact)[^\\s/(;)#|,=!>]{0,2}(?:detail|details)|(?:mobile|phone|cellphone|contact)[^\\s/(;)#|,=!>]{0,10}(?:number|nbr|num|no)|(?:primary|secondary)[^\\s/(;)#|,=!>]{0,10}(phone|landline|telephone))"
```
[the remark this answers is not in the document]
Comment (luizleite-hotmart) [reply rules/sources/contact_data.yaml]: Thanks a lot, I didn't check the cases that you mention. I just update the PR, the suggestion would still get the noisy terms so I need to change to this.


Captura de Tela 2022-08-29 às 08 25 13

File: rules/sources/contact_data.yaml [hunk 2]
```diff
@@ -25,7 +25,7 @@ sources:
     isSensitive: False
     sensitivity: medium
     patterns:
-      - "(?i).*((phone|mobile|cellphone|contact)[^\\s/(;)#|,=!>]{0,2}(detail|details)|(?:mobile|phone|cellphone|contact)[^\\s/(;)#|,=!>]{0,10}(?:number|nbr|num|no)|(?:primary|secondary)[^\\s/(;)#|,=!>]{0,10}(phone|landline|telephone))"
+      - "(?i)(((^(?!\s*$))(phone|mobile|cellphone|contact))|.*(phone|mobile|cellphone|contact)[^\\s/(;)#|,=!>]{0,2}(?:detail|details)|(?:mobile|phone|cellphone|contact)[^\\s/(;)#|,=!>]{0,10}(?:number|nbr|num|no)|(?:primary|secondary)[^\\s/(;)#|,=!>]{0,10}(phone|landline|telephone))"
```
Comment (HiteshMah-Jan) [inline rules/sources/contact_data.yaml:28]: 1. Adding a `(?!$)` will not match with variables named `phone`
2. The updated regex does not match with `employee_phone_number` any more. This seems the result of parentheses being removed from around the entire older regex. The `.*` is now a part of the first alternative `.*(phone|mobile|cellphone|contact)[^\\s/(;)#|,=!>]{0,2}(?:detail|details)`, while it should come before all the alternatives followed something like : `.*((phone|mobile|cellphone|contact)[^\\s/(;)#|,=!>]{0,2}(?:detail|details)|(?:mobile|phone|cellphone|contact)[^\\s/(;)#|,=!>]{0,10}(?:number|nbr|num|no)|(?:primary|secondary)[^\\s/(;)#|,=!>]{0,10}(phone|landline|telephone))`

**Review by HiteshMah-Jan — CHANGES_REQUESTED** [review]

**Comment by ojaswa1942** [comment]
Fixes #43

**Comment by HiteshMah-Jan** [comment]
Hi @luizleite-hotmart,
We would love to have this contribution. Although, the current suggested regex needs a couple of changes :

1. `(^(?!\s*$))`  is not required, since we need to get a full-match only on ‘phone’ keyword.  [refer](https://github.com/Privado-Inc/privado/pull/42#discussion_r957222131)
2. The updated regex does not match with `employee_phone_number` any more. This seems the result of parentheses being removed from around the entire older regex.

Something like this should work:
```
(?i)((phone|mobile|cellphone|contact)|(.*((phone|mobile|cellphone|contact)[^\\s/(;)#|,=!>]{0,2}(?:detail|details)|(?:mobile|phone|cellphone|contact)[^\\s/(;)#|,=!>]{0,10}(?:number|nbr|num|no)|(?:primary|secondary)[^\\s/(;)#|,=!>]{0,10}(phone|landline|telephone))))
```

**Comment by luizleite-hotmart** [comment]
> Hi @luizleite-hotmart, We would love to have this contribution. Although, the current suggested regex needs a couple of changes :
> 
> 1. `(^(?!\s*$))`  is not required, since we need to get a full-match only on ‘phone’ keyword.  [refer](https://github.com/Privado-Inc/privado/pull/42#discussion_r957222131)
> 2. The updated regex does not match with `employee_phone_number` any more. This seems the result of parentheses being removed from around the entire older regex.
> 
> Something like this should work:
> 
> ```
> (?i)((phone|mobile|cellphone|contact)|(.*((phone|mobile|cellphone|contact)[^\\s/(;)#|,=!>]{0,2}(?:detail|details)|(?:mobile|phone|cellphone|contact)[^\\s/(;)#|,=!>]{0,10}(?:number|nbr|num|no)|(?:primary|secondary)[^\\s/(;)#|,=!>]{0,10}(phone|landline|telephone))))
> ```

Thanks a lot for your help I made a feel tests and it seems working with the new code

**Review by HiteshMah-Jan — APPROVED** [review]

**Comment by gnittala** [comment]
Curious about this - can _landline_ be a noun used to identify a phone number?

**Comment by HiteshMah-Jan** [comment]
Hi @gnittala, totally agree.
The current regex fails to match patterns like `landline-number` :)
Feel free to update the regex and play around, we love such contributions made to privado.

**Comment by pandurangpatil** [comment]
@gnittala refer this [documentation](https://docs.privado.ai/getting-started-with-privado/command-reference) for CLI command . `scan` command accepts additional flag `-c, --config` to pass external rules folder. You can clone this repo https://github.com/Privado-Inc/privado/ make changes in the rule files and pass the cloned repo location as a parameter with `-c, --config`.

**Comment by gnittala** [comment]
Thank you both for the response.  @pandurangpatil I agree that this could be the case of an external / custom configuration that the user can pass instead of adding it to the _base_ regex. 
Do you think it will help to include this as an example of when a user can consider using a custom configuration vs. modifying the _base_ regex ?

**Comment by pandurangpatil** [comment]
@gnittala Yea we are working on the documentation which will give more clarity on when to use custom configuration vs modifying the base regex. 

Having said that the larger point I wanted to make here is that. Even in case one need to modify the base regex. One can change those by cloning the given repo, test those changes using `-c, --config` and raise the pull request. We are also working on adding positive and negative test cases. So while updating existing regex it should not break working use cases. 

In the event, that the changes cannot be merged considering the larger number of use cases. Even in that case, the user is not blocked for changes to be merged with base rules. One can keep on using modified/updated rules by passing them as custom config.

**Pull request merged**